esm: allow `resolve` to optionally return import assertions (#46153)
nodejs/node: doc/api/esm.md

    description: Add support for chaining resolve hooks. Each hook must either
      call `nextResolve()` or include a `shortCircuit` property set to `true`
      in its return.
    pr-url: https://github.com/nodejs/node/pull/40250
    description: Add support for import assertions.
-->

> The loaders API is being redesigned. This hook may disappear or its
> signature may change. Do not rely on the API described below.

>   * `importAssertions` {Object} The object after the `assert` in an `import`
>     statement, or the value of the `assert` property in the second argument of
>     an `import()` expression; or an empty object

Reviewer: ```suggestion
  * `importAssertions` {Object} An object whose key+value pairs represent the
    assertions for the module to import
```

Author: I’m not sure this is any better. Can we call this a wrap for now? I’ve pushed a lot of changes.

Reviewer: The current text is technically wrong, which I don't think is acceptable for a technical documentation. I'm not found of my version, but I'd rather leave the description empty that fill it in with something incorrect, so that's also an option if you prefer.

> I’ve pushed a lot of changes.

Yeah what's up with that? It looks like a lot of unrelated changes have made their way to this PR, making the linter unhappy. May suggest to use fixup commits instead of force pushing to simplify reviews and limit the risk of such mistake to slip through?

Author: I don’t know what happened with `esm.md`, it’s gotten destroyed somehow. Working on it.

  Node.js default `resolve` hook after the last user-supplied `resolve` hook
  * `specifier` {string}
  * `context` {Object}
* Returns: {Object}
  * `format` {string | null | undefined} A hint to the load hook (it might be
    ignored)
    `'builtin' | 'commonjs' | 'json' | 'module' | 'wasm'`
  * `importAssertions` {Object | undefined} The import assertions to use when
    caching the module (optional; if excluded the input will be used)
  * `shortCircuit` {undefined | boolean} A signal that this hook intends to
    terminate the chain of `resolve` hooks. **Default:** `false`
  * `url` {string} The absolute URL to which this input resolves

The `resolve` hook chain is responsible for telling Node.js where to find and
how to cache a given `import` statement or expression. It can optionally return
its format (such as `'module'`) as a hint to the `load` hook. If a format is
specified, the `load` hook is ultimately responsible for providing the final
`format` value (and it is free to ignore the hint provided by `resolve`); if
`resolve` provides a `format`, a custom `load` hook is required even if only to
pass the value to the Node.js default `load` hook.

> Import assertions are part of the cache key for saving loaded modules into the

Reviewer: ```suggestion
Import type assertions are part of the cache key for saving loaded modules into the
```

Node.js internal module cache. The `resolve` hook is responsible for returning
an `importAssertions` object if the module should be cached with different
assertions than were present in the source code (for example, if no assertions
were present but the module should be cached with assertions

> `{ type: 'json' }`).

Reviewer 2: I feel like the key thing is an implementation detail. The reason a loader needs to customize import assertions is to support modules that import JSON without the assertion (bypass the validation).

Author: I’m not sure what you want changed, if anything?

Reviewer 2: I don't know how to write it differently. If you think users can understand the current wording that's fine.

* `specifier` {string}
* `context` {Object}
  * `conditions` {string\[]} Export conditions of the relevant `package.json`
  * `importAssertions` {Object} The object after the `assert` in an `import`

>     statement, or the value of the `assert` property in the second argument of

Reviewer: The "thing" after the `assert` keyword in static imports is technically not an object... I guess that's why it was never documented, it's just not easy to describe.

Author: It _is_ an object for `import()`. I don’t know what to call it for an `import` statement, but it looks like an object.

Reviewer: Note that my comment was specifically about static imports.
Regarding dynamic imports, it's also incorrect, we do not pass the value, otherwise that would mean that the loader mutating the object would have an impact on the user code.

I've sent a suggestion in https://github.com/nodejs/node/pull/46153#discussion_r1066435761.

>     an `import()` expression; or an empty object

Reviewer: `parentURL` doc was removed somehow
```suggestion
    an `import()` expression; or an empty object
  * `parentURL` {string|undefined} The module importing this one, or undefined
```

  * `format` {string|null|undefined} A hint to the load hook (it might be
    ignored)
    `'builtin' | 'commonjs' | 'json' | 'module' | 'wasm'`
  * `importAssertions` {Object|undefined} The import assertions to use when
    caching the module (optional; if excluded the input will be used)
  * `shortCircuit` {undefined|boolean} A signal that this hook intends to
    terminate the chain of `resolve` hooks. **Default:** `false`
  * `url` {string} The absolute URL to which this input resolves

The `resolve` hook chain is responsible for telling Node.js where to find and
how to cache a given `import` statement or expression. It can optionally return
its format (such as `'module'`) as a hint to the `load` hook. If a format is
specified, the `load` hook is ultimately responsible for providing the final
`format` value (and it is free to ignore the hint provided by `resolve`); if
`resolve` provides a `format`, a custom `load` hook is required even if only to
pass the value to the Node.js default `load` hook.

> Import type assertions are part of the cache key for saving loaded modules into
> the Node.js internal module cache. The `resolve` hook is responsible for

Reviewer: `the Node.js internal module cache` is kind of unclear, as Node.js has several module caches.
```suggestion
Import type assertions are part of the cache key for saving loaded modules into
the internal module cache. The `resolve` hook is responsible for
```
or
```suggestion
Import type assertions are part of the cache key for saving loaded modules into
the Node.js internal ES module cache. The `resolve` hook is responsible for
```

returning an `importAssertions` object if the module should be cached with

> different assertions than were present in the source code (for example, if no
> assertions were present but the module should be cached with assertions
> `{ type: 'json' }`).

Reviewer: nit: I'm not sure the example is very clear, we might be better off leaving it out, and adding a full working example in the other repo.
```suggestion
different assertions than were present in the source code.
```

  * `context` {Object}
* Returns: {Object}
  * `format` {string}
  * `shortCircuit` {undefined | boolean} A signal that this hook intends to
    terminate the chain of `resolve` hooks. **Default:** `false`
  * `source` {string | ArrayBuffer | TypedArray} The source for Node.js to

>     evaluate

Reviewer 2: do we use spaces in type unions elsewhere in the docs?
